docs/2020-06-17-Checking-Implication.md

```diff
@@ -0,0 +1,161 @@
+# Checking Implication
+
+Checking implication is an important step for proving one-path and all-path reachability claims.
+This document will describe logically how this step will be performed.
+
+## Background
+
+The reader should be familiar with the algorithm for proving one-path and all-path reachability claims:
+
+- [One-Path](2018-11-08-One-Path-Reachability-Proofs.md)
+- [All-Path](2019-03-28-All-Path-Reachability-Proofs.md)
+
+## Problem
+
+Consider a reachability claim,
+
+```
+φ(X) → ∘ ∃ Y. ψ(X, Y)
+```
+
+where `∘` denotes the appropriate one-path or all-path modality
+and `X` and `Y` denote _distinct_ families of variables.
+We will assume that `φ(X)` and `ψ(X, Y)` are function-like patterns (matching at most one value)
+because this is true in typical use.
+
+The reachability proof algorithm will unroll the modality, giving
+
+```
+φ(X) → ∃ Y. ψ(X, Y) ∨ • ∘ ∃ Y. ψ(X, Y)
+```
+
+If the direct implication holds,
+
+```
+φ(X) → ∃ Y. ψ(X, Y)
+```
+
+then the claim is proven.
+Otherwise, the algorithm continues with
+
+```
+φ(X) → • ∘ ∃ Y. ψ(X, Y)
+```
+
+by rewriting `φ(X)`.
+Here, were are concerned with checking the direct implication,
+
+```
+φ(X) → ∃ Y. ψ(X, Y)
+```
+
+or equivalently, checking that
+
+```
+φ(X) ∧ ¬ ∃ Y. ψ(X, Y)
+```
+
+is unsatisfiable.
+
+## Solution
+
+Checking that the formula
+
+```
+φ(X) ∧ ¬ ∃ Y. ψ(X, Y)
+```
+
+is unsatisfiable is equivalent to checking
+
+```
+⌈φ(X) ∧ ¬ ∃ Y. ψ(X, Y)⌉.
+```
+
+Because we assume that `φ(X)` is a function-like pattern,
+we can move the negation outside of `⌈ ⌉`:
+
+```
+⌈φ(X) ∧ ¬ ∃ Y. ψ(X, Y)⌉
+===
+¬ ⌈φ(X) ∧ ∃ Y. ψ(X, Y)⌉ ∧ ⌈φ(X)⌉.
+```
+
+By alpha-equivalence, we can also move the quantifier outside of `⌈ ⌉`:
+
+```
+(¬ ⌈φ(X) ∧ ∃ Y. ψ(X, Y)⌉) ∧ ⌈φ(X)⌉
+===
+(¬ ∃ Y. ⌈φ(X) ∧ ψ(X, Y)⌉) ∧ ⌈φ(X)⌉.
+```
+
+## Optimization
+
+We can optimize this solution the typical case where
+
+```
+φ(X) = t₁(X) ∧ P₁(X)
+ψ(X, Y) = t₂(X, Y) ∧ P₂(X, Y)
+```
+
+where `t₁(X)` and `t₂(X, Y)` are term-like patterns
+and `P₁(X)` and `P₂(X, Y)` are predicates:
+
+```
+\not(∃ Y. ⌈φ(X) ∧ ψ(X, Y)⌉) ∧ ⌈φ(X)⌉
+===
+\not(∃ Y. ⌈t₁(X) ∧ t₂(X, Y)⌉ ∧ P₁(X) ∧ P₂(X, Y)) ∧ ⌈t₁(X)⌉ ∧ P₁(X)
+===
+\not((∃ Y. ⌈t₁(X) ∧ t₂(X, Y)⌉ ∧ P₂(X, Y)) ∧ P₁(X)) ∧ ⌈t₁(X)⌉ ∧ P₁(X)
+===
+\or(\not(∃ Y. ⌈t₁(X) ∧ t₂(X, Y)⌉ ∧ P₂(X, Y)), ¬ P₁(X)) ∧ ⌈t₁(X)⌉ ∧ P₁(X)
+===
+\or(\not(∃ Y. ⌈t₁(X) ∧ t₂(X, Y)⌉ ∧ P₂(X, Y)) ∧ ⌈t₁(X)⌉ ∧ P₁(X), ¬ P₁(X) ∧ ⌈t₁(X)⌉ ∧ P₁(X))
+===
```

Comment:
I found it pretty hard to read these two lines. Maybe it would be clearer if the `\or` and `\not` would be indented in some way?